Create RStudio_user.rst (#950)
datalad-handbook/book: docs/usecases/RStudio_user.rst

The Challenge
^^^^^^^^^^^^^

Max has been using Rstudio together with :term:`GitHub` for a long time. They know how :term:`Git`
works. Max has learned that Git will not work with their new project,
because there will be too many files and some dataset files will be too large.
Max read the DataLad :ref:`Handbook Basics <basics-intro>` and is decided to use DataLad.
They indeed want to version control larger files, and split files in several repositories, linked as "DataLad dataset hierarchies".
Max still want to use Rstudio and a combination of R and python scripts for the
data analysis.

Bobby is a data manager who already learned (the hard way), how to handle datalad
using Rstudio. They have also created a :term:`GIN` repository with :term:`submodule`\s 
for data and for code, using the [Tonic tool and templates](https://gin-tonic.netlify.app).
He is happy to help Max, as he knows this will allow them to record analysis provenance.

Setting up
^^^^^^^^^^

Max follows the handbook and install datalad on his computer.
Max first want to clone the repository on their computer, they use the Rstudio 
`create a new project` function using the SSH address of the parent repository.

> ![Figures of several screenshot demonstrating the creation of new projects on Rstudio](/img/Rstudio-create.jpg)

Author: ```suggestion
.. figure:: /img/Rstudio-create.jpg
   :scale: 80 %
   :alt: screenshot of Rstudio new project creation.

   Figures of several screenshot demonstrating the creation of new projects on Rstudio

![](/img/Rstudio-create.jpg)
```

Bobby start to explain what happened:
Datalad saved the script using Git-annex.
This means that the file was moved somewhere else, and the content was replaced by a code linking to the file location. 
The code, which is a tiny file, is saved in Git, while the large file is saved outside of Git.
Because it is :term:`symlink`, Rstudio still read the content of the original file when clicking on it, but it cannot overwrite the file: that file is in read-only mode.
This is explained in detail in the :ref:`Handbook chapters on Git-annex <basics-annex>`_ .

One could overwrite the file by first unlocking it (using ``datalad unlock .``), but that would not be very practical, and it would save the script as a binary file, making the version control very inefficient.

You do not want to use Git-annex for scripts, as they are text files which version should be handled by Git..
Bobby then shows how to tell datalad to use git for text files and he runs: ``datalad create -c text2git --force``. 

Max can now work on its script as he used to, but commit changes using the ``datalad save -r`` command.

.. gitusernote:: Dangers of text2git

>   Note that all text files will be added to git using this option, so if you have large text files (.csv or .json files) that you want to be added via Git-annex,
> you will need to be more precise in what text file should not be annexed.
> See :ref:`Handbook chapters <101-124-procedures>` ,<http://handbook.datalad.org/en/inm7/basics/101-124-procedures.html#>
> for details on how text2git change `.gitattributes` to achieve that.

Author: ```suggestion
  Note that all text files will be added to git using this option, so if you have large text files (.csv or .json files) that you want to be added via Git-annex, you will need to be more precise in what text file should not be annexed. See :ref:`Handbook chapters <101-124-procedures>` , <http://handbook.datalad.org/en/inm7/basics/101-124-procedures.html#> for details on how text2git change `.gitattributes` to achieve that.
```

.. \_usecase_Rstat:

DataLad and RStudio: First steps
---------------------------

.. index:: ! Usecase; R users quickstart

This use case sketches typical entry points for R and `Rstudio <https://en.wikipedia.org/wiki/RStudio>`_ users. 

#. A repository having submodules for data and code is cloned.
#. R scripts are developed in Rstudio
#. R scripts are run using ``datalad run``

(This is a `hello world` type of analysis, used only for demonstration purposes.)

The Challenge
^^^^^^^^^^^^^

Max has been using Rstudio together with :term:`GitHub` for a long time. They know how :term:`Git`
works. Max has learned that Git will not work with their new project,
because there will be too many files and some dataset files will be too large.

> Max read the DataLad :ref:`Handbook Basics <basics-intro>`_ and is decided to use DataLad.

Author: ```suggestion
Max read the DataLad :ref:`Handbook Basics <basics-intro>` and is decided to use DataLad.
```

Bobby start to explain what happened:
Datalad saved the script using Git-annex.
This means that the file was moved somewhere else, and the content was replaced by a code linking to the file location. 
The code, which is a tiny file, is saved in Git, while the large file is saved outside of Git.
Because it is :term:`symlink`, Rstudio still read the content of the original file when clicking on it, but it cannot overwrite the file: that file is in read-only mode.
This is explained in detail in the :ref:`Handbook chapters on Git-annex <basics-annex>`_ .

One could overwrite the file by first unlocking it (using ``datalad unlock .``), but that would not be very practical, and it would save the script as a binary file, making the version control very inefficient.

You do not want to use Git-annex for scripts, as they are text files which version should be handled by Git..
Bobby then shows how to tell datalad to use git for text files and he runs: ``datalad create -c text2git --force``. 

Max can now work on its script as he used to, but commit changes using the ``datalad save -r`` command.

.. gitusernote:: Dangers of text2git

>   Note that all text
> files will be added to git using this option, so if you have large text files
> (.csv or .json files) that you want to be added via Git-annex,
> you will need to be more precise in what text
> file should not be annexed.
> See :ref:`Handbook chapters <101-124-procedures>`_ <http://handbook.datalad.org/en/inm7/basics/101-124-procedures.html#>
> for details on how text2git change `.gitattributes` to achieve that.

Author: ```suggestion
  Note that all text files will be added to git using this option, so if you have large text files (.csv or .json files) that you want to be added via Git-annex,
you will need to be more precise in what text file should not be annexed.
See :ref:`Handbook chapters <101-124-procedures>` ,<http://handbook.datalad.org/en/inm7/basics/101-124-procedures.html#>
for details on how text2git change `.gitattributes` to achieve that.
```

---------------------------

.. index:: ! Usecase; R users quickstart

This use case sketches typical entry points for R and `Rstudio <https://en.wikipedia.org/wiki/RStudio>`_ users. 

#. A repository having submodules for data and code is cloned.
#. R scripts are developed in Rstudio
#. R scripts are run using ``datalad run``

(This is a `hello world` type of analysis, used only for demonstration purposes.)

The Challenge
^^^^^^^^^^^^^

Max has been using Rstudio together with :term:`GitHub` for a long time. They know how :term:`Git`
works. Max has learned that Git will not work with their new project,
because there will be too many files and some dataset files will be too large.
Max read the DataLad :ref:`Handbook Basics <basics-intro>` and is decided to use DataLad.
They indeed want to version control larger files, and split files in several repositories, linked as "DataLad dataset hierarchies".
Max still want to use Rstudio and a combination of R and python scripts for the
data analysis.

Bobby is a data manager who already learned (the hard way), how to handle datalad
using Rstudio. They have also created a :term:`GIN` repository with :term:`submodule`\s 

> for data and for code, using the [Tonic tool and templates](https://gin-tonic.netlify.app).

Reviewer: This statement is primarily emotion (which is OK). However, instead of merely stating that there is also a "GIN" and a "Tonic" part of the scenario, it would be better to say "why".

GIN likely is here, because it could host large files. If that is the is main reason, it should make that explicit. If so, it would be useful to say that GIN is one convenient solution, but not the only one.

Likewise the role of Tonic should be clarified. Right now it seems that it is needed to set up the GIN repos.

Given that the main content of this use case is about executing code locally (and not about pushing the results somehere), it is worth considering moving this information into a box, to avoid the impression that it is necessary to have these for running RStudio with DataLad.

``datalad create --force -r`` in the parent repository.
Now they are sure they set up datalad to work in the repository and all submodules,
since they used the ``-r``  option.

  
Working on the code
^^^^^^^^^^^^^^^^^^^

Max starts to write some script he saves in the analysis submodule, and use :command:`datalad save` typing ``datalad save -r -m "this is a first draft of the script"`` command in the terminal (in the parent repository). 
The commit history of the parent and the analysis repositories shows the message and Max things everything works fine.
Max change the script, but Rstudio does not want to save the changes.
Max save a copy of the script file and call Bobby for help.

Bobby start to explain what happened:
Datalad saved the script using Git-annex.
This means that the file was moved somewhere else, and the content was replaced by a code linking to the file location. 
The code, which is a tiny file, is saved in Git, while the large file is saved outside of Git.
Because it is :term:`symlink`, Rstudio still read the content of the original file when clicking on it, but it cannot overwrite the file: that file is in read-only mode.

> This is explained in detail in the :ref:`Handbook chapters on Git-annex <basics-annex>`_ .

Author: ```suggestion
This is explained in detail in the :ref:`Handbook chapters on Git-annex <basics-annex>`.
```

Running code
^^^^^^^^^^^^

The code use relative path starting in the parent repository, as they are used to do in normal projects, and since the code is run from there in Rstudio.
 (Later on, Max realise he can also use git commands from inside the analysis submodule, and he creates a second Rstudio project in that submodule, just to use the git functions he is used to. Code is run from the parent Rstudio project.)

Max is now happy and start working on his code. 
In order to test everything, Max put a text file in the data submodule, and write a script that read the file and produce a pdf writing the text as an image.
He runs the code and it works!
He know save it with ``datalad save -r``.
He runs the code again and... oups it fails.

Max thinks a bit about it and remember what he learned before: the pdf file has been annexed and cannot be overwritten.
Max therefore runs ``datalad unlock . -r`` and then runs the code, and it works.

> Max realise also that usinng ``datalad save . -r`` lock the files again, 

Author: ```suggestion
Max realise also that using ``datalad save . -r`` lock the files again, 
```

The path is relative to where the terminal is, the terminal tab is per default in the working directory of the project. If your code is in a submodule and the data is in another one, you should run this command from the parent repository.

(Bobby needs here to make sure Rscript is a recognised command and set the PATH variable accordingly.)

"What are the advantages of using this command", asks Max.

There are twofolds (at least), answers Bobby.
First, this command will take care of obtaining input files and unlocking output files for you.
Second, and most importantly, the command will record what has been done in the commit message automatically: what input, what script, what output was used.
This code therefore records **provenance**: you will always be able to find what workflow and data version was used to create your figures.

Since Bobby looks very enthusiastic about provenance, Max reads a little more about it in the handbook: usecases/provenance_tracking, https://handbook.datalad.org/en/latest/basics/101-108-run.html#run

Then, Max create a bash script in Rstudio and run it using the usual button (This runs the bash script in the terminal).

> .. code-block:: bash
> 
>     
>     $ datalad run  \
>       --input "file1.csv" \
>       --input "data/file2.json" \
>       --output "figures/*.png" \
>       --explicit \

Author: ```suggestion
.. code-block:: bash

    
$ datalad run  \
--input "file1.csv" \
--input "data/file2.json" \
--output "figures/*.png" \
--explicit \
Rscript "<path-to-script.r>" {inputs} {outputs}
```

> $ datalad run  \
> --input "file1.csv" \
> --input "data/file2.json" \
> --output "figures/*.png" \
> --explicit \
> Rscript "<path-to-script.r>" {inputs} {outputs}

Author: ```suggestion
  $ datalad run  \
  --input "file1.csv" \
  --input "data/file2.json" \
  --output "figures/*.png" \
  --explicit \
  Rscript "<path-to-script.r>" {inputs} {outputs}
```

>     $ datalad run  \
>       --input "file1.csv" \
>       --input "data/file2.json" \
>       --output "figures/*.png" \
>       --explicit \
>       Rscript "<path-to-script.r>" {inputs} {outputs}

Author: ```suggestion
$datalad run  \
--input "file1.csv" \
--input "data/file2.json" \
--output "figures/*.png" \
--explicit \
Rscript "<path-to-script.r>" {inputs} {outputs}
```

.. figure:: ../artwork/src/Rstudio/Rstudio-dataladrun.jpg
   :scale: 80 %
   :alt: screenshot of Rstudio bash code window

   Figures of bash code running the datalad run command

> On can set as many input and output files, one can use `*` to define several files with a similar ending (in the example all .png figures will be unlocked). It is good practice to list files in input and output even if they do not need to be handled by datalad, in order to give more information in the commit message.

Reviewer: ```suggestion
One can set as many input and output files, one can use `*` to define several files with a similar ending (in the example all .png figures will be unlocked). It is good practice to list files in input and output even if they do not need to be handled by DataLad, in order to give more information in the commit message.
```

> .. gitusernote:: behavior explained

Reviewer: ```suggestion
.. gitusernote:: Behavior explained
```

- explicit: datalad runs normally only in clean repositories, this includes all submodules. By adding --explicit, datalad will only test that the output files are clean, and only output files will be saved. Please use with care, as the script and data you use will not be tested and provenance information can be lost.
- {inputs} {outputs}: If you add the placeholders, the terminal will actually gives the input and output text as argument to the Rscript bash function. One can access them in the R script with `args <- commandArgs(trailingOnly = TRUE)` (then get them with `args[i]`, with i starts at 1).
- At the end, datalad usually runs `datalad save -r` so that modification made by the code in the whole repository, including submodules will be done (exception when --explicit is given, see above.) This will include any intermediate file created by your code in bash mode, that is using `Rscript "path-to-code.R "` in the terminal (it can happen that bash mode creates more files than running the code directly)  

> .. gitusernote:: advanced tips for datalad run 

Reviewer: ```suggestion
.. gitusernote:: Advanced tips for ``datalad run`` 
```

.. \_usecase_Rstat:

My first steps as a RStudio users
---------------------------

.. index:: ! Usecase; R users quickstart

This use case sketches typical entry points for R and Rstudio users. 

#. A repository having submodules for data and code is cloned.
#. R scripts are developped in Rstudio
#. R scripts are run usind datalad run

This is a `hello world` type of analysis, used only for demonstration purposes

The Challenge
^^^^^^^^^^^^^

Max has been using Rstudio together with GitHub for a long time. They know how git
works. Max has learned that git will not work with their new project,
because there will be too many files and some dataset files will be too large.
Max read the datalad handbook basics and is decided to apply datalad.
Max still want to use Rstudio and a combination of R and python scripts for the
data analysis.

Bobby is a data manager who already learned (the hard way), how to handle datalad
using Rstudio. They have also created a GIN repository with submodules 

> for data and for code, using the Tonic tool and templates.

Author: ```suggestion
for data and for code, using the [Tonic tool and templates](https://gin-tonic.netlify.app).
```

Setting up
^^^^^^^^^^

Max first want to clone the repository on their computer, they use the Rstudio 
`create a new project` function using the SSH address of the parent repository.
Max can't see submodules content and come to Bobby.

Bobby comes and run `datalad get . -n -r` in the terminal window of Rstudio. 

They then explain:
  - Rstudio can only use simple git commands, which do not clone submodule content.
  - datalad command are run in the terminal window. Datalad does not have a R package and do not run in the console
  - This specific function `get .` will download all files, it has two options:
     - `-n` option means annexed files will not be downloaded
     - `-r` option means that the function is run in all submodules, recursively

Max has learned something, and it realise that git-annex is probably not set in its repositoy.
After reading the handbook, they see the create function need to be forced 
when the folder already exist, so they  run 
`datalad create --force -r` in the parent repository.
Now they are sure they set up datalad to work in the repository and all submodules,
since they used the `-r` option.

> 

Reviewer 2: I feel like this can conclude with a statement that with this setup, everything is good to go for DataLad commands from the console, for example for saving changes,pushing modifications, pulling updates, or adding siblings.

Author: Added some infos in the box, indeed pushing would need additional setups in the scenario

Working on the code
^^^^^^^^^^^^^^^^^^^

> 

Reviewer 2: Maybe for this section it makes sense to focus it on the topic of reproducible execution with datalad run.
I think there is no need to spent too much work on rewriting content about the difference between files kept in git versus in git annex (instead, references to existing parts in the handbook).

Author: I kept this in the following version because:
- beginners may not care about reproducible execution
- Infos about difference between Git and Git-annex is necessary to explain Rstudio behavior. I was personally very surprised of that behavior and needed testing and thinking to understand what happens. I try to add some reference for more info.

> .. \_usecase_Rstat:

Reviewer 2: ```suggestion
.. _usecase_Rstat:
```

> DataLad and RStudio: First steps
> ---------------------------

Reviewer 2: ```suggestion
DataLad and RStudio: First steps
--------------------------------
```

Max is a bit puzzled and read the basics chapter of the handbook again.
Then, they see that :command:`datalad create --force` is the correct command  to create a DataLad dataset.
when the folder already exist, so they  run 
``datalad create --force -r`` in the parent repository.
Now they are sure they set up DataLad to work in the repository and all submodules,
since they used the ``-r``  option.
Max also now understand what Bobby meant when he said that annexed files are not downloaded, and this is further explained in the next section.

.. gitusernote:: Pushing and GIN-Tonic

  The scenario  does not present the use of remotes (server version of the resitory) and :command:`datalad push`, as it is indeed not different for R or Rstudio users.
  One *only* needs to use DataLad commands (in the terminal tab) instead of Git (integration in Rstudio-GUI or shell).
  
  While the original repositories are described as pure Git repositories here,
  most of the scenario was actually tested
  using a :term:`GIN` repository with git-annex, with :term:`submodule`\s 

>   for data and for code, that were created using the `Tonic tool and templates <https://gin-tonic.netlify.app>`_ . 

Author: ```suggestion
  for data and for code, that were created using the `Tonic tool and templates <https://gin-tonic.netlify.app>`_ , `see archived version <https://doi.org/10.5281/zenodo.7763694>`_
```

Author: adding a link, so that if ever the website is put down, there is still a working link.

  Note that the template itself `was ported from a single GitHub repository into DataLad datasets <https://gin-tonic.netlify.app/installation/template_repository/>`_ . 
  
  
  
    
Working on the code
^^^^^^^^^^^^^^^^^^^

Max starts to write one script that he saves in the analysis submodule, and use :command:`datalad save` typing ``datalad save -r -m "this is a first draft of the script"`` command in the terminal (in the parent repository). 
The commit history of the parent and the analysis repositories shows the message and Max things everything works fine.
Max change the script, but Rstudio does not want to save the changes.
Max save a copy of the script file and call Bobby for help.

Bobby start to explain what happened:
DataLad saved the script using Git-annex.
This means that the file was moved somewhere else, and the content was replaced by a code linking to the file location. 
Because it is :term:`symlink`, Rstudio still read the content of the original file when clicking on it, but it cannot overwrite the file: that file is in read-only mode.

> This is explained in detail in the :ref:`Handbook chapters on Git-annex <basics-annex>`.

Reviewer 2: ```suggestion
This is explained in detail in the :ref:`Handbook chapters on Git-annex <chapter_gitannex>`.
```

One could overwrite the file by first unlocking it (using ``datalad unlock .``), but that would not be very practical, and it would save the script as a binary file, making the version control very inefficient.
You do not want to use Git-annex for scripts, as they are text files which version should be handled by Git.
Bobby then shows how to tell DataLad to use git for text files and he runs: ``datalad create -c text2git --force``. 

Max can now work on its script as he used to, but commit changes using the ``datalad save -r`` command.

.. gitusernote:: The dangers of text2git

  Note that all text files will be added to git using this option, so if you have large text files (.csv or .json files) that you want to be added via Git-annex, you will need to be more precise in what text file should not be annexed. 
  This may be quite important if you want to be able to *drop* files to keep space on your computer harddrive.

>   See :ref:`Handbook chapters <basics-101-124-procedures>` for details on how text2git change `.gitattributes` to achieve that.

Reviewer 2: ```suggestion
  See :ref:`Handbook chapters <procedures>` for details on how text2git change `.gitattributes` to achieve that.
```

Then, Max create a bash script in Rstudio and run it using the usual button (This runs the bash script in the terminal). After some practice and tests,
Max finds his way between testing/developing code on training data (as he has been used to, while using :command:`datalad unlock` and :command:`datalad get` when needed), and running ready code on real data in a reproducible way using :command:`datalad run`. 

 

.. code-block:: bash

    
  $ datalad run  \
  --input "file1.csv" \
  --input "data/file2.json" \
  --output "figures/*.png" \
  --explicit \
  Rscript "<path-to-script.r>" {inputs} {outputs}

.. figure:: ../artwork/src/rstudio/Rstudio-dataladrun.png
   :scale: 80 %
   :alt: screenshot of Rstudio bash code window

   Figures of bash code running the DataLad run command

One can set as many input and output files, one can use `*` to define several files with a similar ending (in the example all .png figures will be unlocked). It is good practice to list files in input and output even if they do not need to be handled by DataLad, in order to give more information in the commit message.

.. gitusernote:: Behavior explained

> - Input: To be read, files are downloaded if not present. Note that they are not unlocked (no need for reading them) and that they will not be dropped again after being read.
> - Output: files are unlocked so they can be overwritten. If the files are not present (dropped), they will not be downloaded. This may make your code fail: if it does, either get the files manually before running `datalad run`, or remove them in the R code (`r file.remove()`). In other case, it will work and it will even detect when the file has not been modified and make no commit.
> - explicit: :command:`datalad run` runs normally only in clean repositories, this includes all submodules. By adding --explicit, DataLad will only test that the output files are clean, and only output files will be saved. Please use with care, as the script and data you use will not be tested and provenance information can be lost.
> - {inputs} {outputs}: If you add the placeholders, the terminal will actually gives the input and output text as argument to the Rscript bash function. One can access them in the R script with `args <- commandArgs(trailingOnly = TRUE)` (then get them with `args[i]`, with i starts at 1).
> - At the end, DataLad usually runs `datalad save -r` so that modification made by the code in the whole repository, including submodules will be recorded and committed (exception when --explicit is given, see above.) This will include any intermediate file created by your code in bash mode, that is using `Rscript "path-to-code.R "` in the terminal (it can happen that bash mode creates more files than running the code directly).  

Author: ```suggestion
  - Input: To be read, files are downloaded if not present. Note that they are not unlocked (no need for reading them) and that they will not be dropped again after being read.
  - Output: files are unlocked so they can be overwritten. If the files are not present (dropped), they will not be downloaded. This may make your code fail: if it does, either get the files manually before running `datalad run`, or remove them in the R code (`r file.remove()`). In other case, it will work and it will even detect when the file has not been modified and make no commit.
  - explicit: :command:`datalad run` runs normally only in clean repositories, this includes all submodules. By adding --explicit, DataLad will only test that the output files are clean, and only output files will be saved. Please use with care, as the script and data you use will not be tested and provenance information can be lost.
  - {inputs} {outputs}: If you add the placeholders, the terminal will actually gives the input and output text as argument to the Rscript bash function. One can access them in the R script with `args <- commandArgs(trailingOnly = TRUE)` (then get them with `args[i]`, with i starts at 1).
  - At the end, DataLad usually runs `datalad save -r` so that modification made by the code in the whole repository, including submodules will be recorded and committed (exception when --explicit is given, see above.) This will include any intermediate file created by your code in bash mode, that is using `Rscript "path-to-code.R "` in the terminal (it can happen that bash mode creates more files than running the code directly).  
```

> .. gitusernote:: Advanced tips for ``datalad run`` 

Author: ```suggestion
.. gitusernote:: Advanced tips for the DataLad run command
```

Author: the `` thing does not work in note title.

  unlocking the files will make its state "unclean", so if you use DataLad run, you need to set output options in the function, you cannot unlock files manually before.

  The commit message will only look at the options, whether the code use these input and output files is not checked.
   
  Using :command:`datalad run` correctly is sometimes tricky, and since it does save each time, it can make the repository history quite messy. Make sure to give good commit messages by adding a `-m` variable to it. 

.. importantnote:: Take home messages
  
  DataLad commands run in the terminal, not the R Console.
  
  The simplest way to tell DataLad not to use git-annex for your code files is to use ``datalad create -r -c text2git --force`` command.

  the ``datalad run Rscript "path-to-script.r"`` command will run your script.
  

>   Use additional options to command:`datalad run` to read or write annexed files (and give more info for commit messages).

Author: ```suggestion
  Use additional options to command: `datalad run` to read or write annexed files (and give more info for commit messages).
```

  In your R script, use path relative to the project, not relative to the code position.
  

>   An understanding of how Git-annex works and why, makes it easier to understand the workflow, refer to :ref:`Handbook Git-annex Basics <basics-annes>`.  

Reviewer 2: ```suggestion
  An understanding of how Git-annex works and why, makes it easier to understand the workflow, refer to :ref:`Handbook Git-annex Basics <chapter_gitannex>`.  
```
